### ISSUE TITLE
Forced semver ends up with two different versions when patch is used

### ISSUE THREAD

**Issue by illicitonion** [rust-lang/cargo#7107]
**Problem**
I have forked a crate (protobuf==2.0.6) which I use in my crate. I have an explicit dep on it, and I have a patch section which points at my fork.
I have a dependency (protoc-grpcio==0.2) which itself has a dep on protobuf ^2

When I ran `cargo update`, I found that I now have two copies of protobuf, my fork at 2.0.6, and another at 2.7.0 depended on by protoc-grpcio.

I expect to have exactly one version 2.0.6, because that satisfies the transitive dependency constraint (^2). Instead, I have two, which causes compilation errors because of type mismatches.

If I remove my patch section, I end up with exactly 2.0.6, and no transitive dep on 2.7, as expected.

How can I both have my patch section take effect, and 

**Steps**
With this Cargo.toml:
```
[package]
version = "0.0.1"
edition = "2018"
name = "root"

[dependencies]
protobuf = { version = "=2.0.6" }
protoc-grpcio = "=0.2"

[patch.crates-io]
protobuf = { git = "https://github.com/pantsbuild/rust-protobuf", rev = "171611c33ec92f07e1b7107327f6d0139a7afebf", version = "=2.0.6" }
```

run:
`cargo update`
Expect `grep 'protobuf 2\.7' Cargo.lock` to return nothing. Instead, find there's a dep on protobuf 2.7.

**Comment by alexcrichton** [comment]
Thanks for the report! Using `[patch]`, however, simply adds new candidates for resolution and there's no guarantee that they'll be picked. If you want explicit control over versions you'll need to use `cargo update --precise`, otherwise if 2.7 is a valid solution for the lock file then that will get picked since it's a higher vesion than 2.0.6.

**Comment by illicitonion** [comment]
Thanks for the quick reply! Just to check I understand the behaviour - I had assumed that cargo would attempt to resolve dependencies such that exactly one version of a library is used if possible. So if a few crates have dependencies:

foo = 1
foo = 1.1
foo = 1.1.5

all three of those crates would end up with something that satisfies all three constraints, because that's possible, rather than having two or three different versions of foo linked into the crate.

Is that not something cargo attempts to do?

**Comment by illicitonion** [comment]
Actually, I just tried running `cargo update --precise=protobuf==2.0.6` and I still ended up with two different versions of protobuf linked into my crate; 2.0.6 and 2.7.0... Am I missing something important about the flag?

**Comment by alexcrichton** [comment]
You can read up on the dependency syntax at https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html, but "1", "1.1", and "1.1.5" basically just means ">= 1.1.5, < 2.0.0". All three crates will get the same version.

Cargo only allows one version within a major version track, so it's not possible to have a crate graph with both 2.0.6 and 2.7.0. It may be the case that some output was misinterpreted? You won't be able to downgrade to 2.0.6 if something requires greater than 2.0.6.

The syntax for this would be `cargo update -p protobuf --precise 2.0.6`

**Comment by ehuss** [comment]
@alexcrichton It is allowing both 2.0.6 and 2.7.0 at the same time.  I noticed almost the exact same situation in #6762 when using a `~` dependency (which coincidentally? also involves protobuf).

Using the sample `Cargo.toml` above, running `cargo generate-lockfile` I get:

```
root
    protobuf =2.0.6 (2.0.6)
    protoc-grpcio =0.2 (0.2.1)

protoc-grpcio 0.2.1
    protobuf ^2.0 (2.7.0)
    protoc ^2.0 (2.7.0)
    protobuf-codegen ^2.0 (2.0.6)
```

Notice how there are now two versions of `protobuf`.

What's interesting is if you comment out the `[patch]`, delete `Cargo.lock`, `generate-lockfile`, then you have only one version (2.0.6).  Then uncomment the `[patch]`, and run a command like `cargo check` or `cargo build`, it will pick up the patch without adding the secondary 2.7.0 dependency.

The suggestions I gave in #6762 should also apply here.  The command to run to "fix" it is `cargo update -p protobuf:2.7.0 --precise 2.0.6` (note the colon package spec, needed because there are two versions in the lock file).

I don't know enough about the resolver to speculate why this happens.  Somehow `[patch]` causes transitive dependencies to "float" to the newest version if they are not previously locked via `Cargo.lock`.   A bare `cargo update` will also cause them to be updated to the latest (non-patched) version.

This doesn't seem like correct `[patch]` behavior to me.  Maybe `[patch]` was not designed to downgrade versions?  Or is this a bug?

**Comment by illicitonion** [comment]
> The suggestions I gave in #6762 should also apply here. The command to run to "fix" it is `cargo update -p protobuf:2.7.0 --precise 2.0.6` (note the colon package spec, needed because there are two versions in the lock file).

This worked for me - thanks for the help!

> I don't know enough about the resolver to speculate why this happens. Somehow `[patch]` causes transitive dependencies to "float" to the newest version if they are not previously locked via `Cargo.lock`. A bare `cargo update` will also cause them to be updated to the latest (non-patched) version.
> 
> This doesn't seem like correct `[patch]` behavior to me. Maybe `[patch]` was not designed to downgrade versions? Or is this a bug?

It feels like a bug that there's nothing I can put in my `Cargo.toml` to guarantee a consistent `Cargo.lock` resolution given a valid semver constraint (`protobuf==2.0.6`).

If this _is_ working as intended, it feels like there's a missing `Cargo.toml` section which is "precise up/downgrades to apply when updating by default", but that feels redundant with the dependency semver specification that seems to just be ignored in this case...

**Comment by alexcrichton** [comment]
Ah thanks for pointing that out @ehuss, I've opened https://github.com/rust-lang/cargo/issues/7117 to track this failure mode.
